### PR TITLE
Feature/attach file to email transport builder

### PR DESCRIPTION
**Pull request by enrico69**
Added a new method to the transport build to make joining a file to an email easier.

### Description
Added a method joinFile() to the class Magento\Framework\Mail\Template\TransportBuilder to add some abstraction and making easier to join a file from the transport builder.

### Manual testing scenarios

An example of testing is to send the user agreement in a pdf file joined to the invoice. 
Override the class Magento\Sales\Model\Order\Email\SenderBuilder and call, at line 52, this code:
`$this->transportBuilder->joinFile($directory, $fileName, $fileType, $joinedFileName);` like this
`$this->transportBuilder->joinFile('var', 'EULA.pdf', 'application/pdf', 'EULA.pdf');`

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: I think would be better to:
1. Rename method to attachFile
2. Combine directory & filename to one parameter - filepath
3. Make auto-filling of file type (or define some default value)
4. Make auto-filling of attacked file (by default should be the same)

Comment (enrico69) [reply lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: Ok thank you for your opinion. I am going to make these modifications today.

Comment (ihor-sviziev) [reply lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: I think you could wait for answer from magento team, maybe they decide to do another changes.

Comment (korostii) [reply lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: You almost made my head explode right there.
I think you've meant to write "attachFile", haven't you?

**Comment by fooman** [comment]
@enrico69 thanks for your pull request.

One issue which would need addressing as part of this pull request is that the current 
`Magento\Framework\Mail\MessageInterface`
does not include the method createAttachment(). It currently works because the class that is instantiated `\Zend_Mail` does have it - it however would break on implementations that fulfil the interface but do not have a createAttachment method with the same parameters. 

As  `Magento\Framework\Mail\MessageInterface` is part of the `@api` we will either need to bump it's version number and/or create a new enhanced MessageInterface.

I'll get someone from the Magento side to chime in on the best way forward.

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php
Comment (ihor-sviziev) [reply lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: @korostii yes, fixed my typo

**Comment by fooman** [comment]
Okay so the way to approach the introduction of this new interface is outlined [here](http://devdocs.magento.com/guides/v2.0/contributor-guide/backward-compatible-development/#introduction-of-a-method-to-a-class-or-interface).

**Comment by enrico69** [comment]
Modifications done.

File: lib/internal/Magento/Framework/Mail/ExtendedMessageInterface.php
Comment (vkublytskyi) [inline lib/internal/Magento/Framework/Mail/ExtendedMessageInterface.php]: The name of the interface is too abstract. It may cause an issue if in future new methods should be added to Message. Please reflect purpose of the interface in a name (e.g. `AttachmentMessageInterface`)

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php
Comment (vkublytskyi) [inline lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: This change is not allowed as is backward incompatible. If some 3rd party module has own implementation of MessageInterface it will not be accepted by this class anymore.

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php
Comment (vkublytskyi) [inline lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: Unnecessary changes. Constants are defined at `Magento\Framework\Mail\MessageInterface` so it's better to use more high-level abstraction if possible.

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php
Comment (vkublytskyi) [inline lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: Avoid direct class creation with `new` operator in Magento 2. Use `Magento\Framework\Filesystem\Directory\ReadFactory::create` instead.

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php
Comment (vkublytskyi) [inline lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: `Magento\Framework\Mail\Template\TransportBuilder` is also marked with an `@api` annotation so we should avoid changing its contract.
Probably the best approach to implement this feature would implement a decorator for `Magento\Framework\Mail\Template\TransportBuilder`:
1. Create new class that `extends` `Magento\Framework\Mail\Template\TransportBuilder` (Magento 2 not recommends use class inheritance but in this case seems this is most suitable approach). Class name should reflects class purpose (e.g. `AttachmentTransportBuilder`)
2. This class should accept an instance of `Magento\Framework\Mail\Template\TransportBuilder` in a constructor and delegate all methods calls to it.
3. Define `attachFile` method in the decorator. Instead of modifying own `message` property work directly with a property of decorated object (`$this->decorated->message->createAttachement`). It will be accessible as is `protected` and decorator is a child of decorated class.
4. You may add logic in the constructor and check that message in the decorated object is an instance of message with attachment and throws `\InvalidArgumentException` otherwise.

File: lib/internal/Magento/Framework/Mail/Transport.php
Comment (vkublytskyi) [inline lib/internal/Magento/Framework/Mail/Transport.php]: Modification of constructor is not allowed as this is backward incompatible change. Please see http://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/. Change of argument type to more strict is equal to removing all argument and introduction of new one.

**Review by vkublytskyi — CHANGES_REQUESTED** [review]
@enrico69 Thank you for this contribution and sorry for a long time without response.

Possibility to attach a file to a mail message is really needed feature. However, as these changes modify existing API and introduce new one, such changes should be designed more carefully to not introduce backward incompatible changes and to have a usable API as a result. Please see my comments regarding implementation.

If you are still interested in this PR please resolve backward compatibility issues and  conflicts with changed Magento code.

**Comment by enrico69** [comment]
Ok I am going to check that.

**Comment by enrico69** [comment]
In progress

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=9747) <br/>All committers have signed the CLA.

**Comment by ihor-sviziev** [comment]
@enrico69 let us know once you finish it or if you need any help

**Comment by enrico69** [comment]
Hi @ihor-sviziev 
Sorry, the job is (supposed to be) done in the commit [6cafef1](https://github.com/magento/magento2/pull/9747/commits/6cafef1e5e18d86f1badd83ed7052d22f6fccb34) .

I tried to follow your recommandations as much as possible.

**Comment by okorshenko** [comment]
Hi @enrico69 
could you please cleanup the branch so that we can proceed with this PR.

**Comment by enrico69** [comment]
Ho @okobchenko 

Honestly I am not sure what do now. I did my modifications after trying to update my branch which was veryyy absolete. All my work is in [6cafef1](https://github.com/magento/magento2/pull/9747/commits/6cafef1e5e18d86f1badd83ed7052d22f6fccb34)

**Comment by okorshenko** [comment]
Hi @enrico69 
Please checkout 2.3-develop branch from magento/magento2 repository. Apply the changes from your commit. Please, do not use `root` as a an author of commit. Use your github account. Force push your local changes to `enrico69:feature/join-file-method-transport-builder`

**Comment by enrico69** [comment]
Hi @okorshenko 
OK I will do that asap.

**Comment by enrico69** [comment]
Hi @okorshenko 

That's done.

**Comment by magento-engcom-team** [comment]
@enrico69 thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Review by vkublytskyi — APPROVED** [review]

File: lib/internal/Magento/Framework/Mail/Template/AttachmentTransportBuilder.php
Comment (vkublytskyi) [inline lib/internal/Magento/Framework/Mail/Template/AttachmentTransportBuilder.php:19]: Better to declare as private properties

**Comment by magento-engcom-team** [comment]
Hi @vkublytskyi, thank you for the review.
[MAGETWO-70492](https://jira.corp.magento.com/browse/MAGETWO-70492) has been created to process this Pull Request

File: lib/internal/Magento/Framework/Mail/Template/AttachmentTransportBuilder.php
Comment (TomashKhamlai) [inline lib/internal/Magento/Framework/Mail/Template/AttachmentTransportBuilder.php:3]: Please correct the copyright. Tests failed with error:
`Copyright is missing or has wrong format`

File: lib/internal/Magento/Framework/Mail/AttachmentMessageInterface.php
Comment (TomashKhamlai) [inline lib/internal/Magento/Framework/Mail/AttachmentMessageInterface.php:3]: Please correct the copyright. Tests failed with error:
`Copyright is missing or has wrong format`

**Comment by sidolov** [comment]
Hi @enrico69 , I verified your solution and found that we updated `zendframework/zend-mail` package to the new version. Unfortunately your implementation not compatible with latest changes. You may implement a new solution or close this PR. 
Thank you for your contribution!

**Comment by sidolov** [comment]
Hi @enrico69 , I'm closing this PR for now, if you implement the new solution, feel free to reopen this PR or create new. Thank you for collaboration!

**Pull request closed without merging**
